From 83640cb1a20b065bd53f20f42322d631f340663f Mon Sep 17 00:00:00 2001
From: Philippe Gerum <rpm@xenomai.org>
Date: Sat, 9 Apr 2022 16:19:51 +0200
Subject: [PATCH 085/217] x86/fpu: dovetail: enable alternate scheduling
MIME-Version: 1.0
Content-Type: text/plain; charset=UTF-8
Content-Transfer-Encoding: 8bit

Signed-off-by: Philippe Gerum <rpm@xenomai.org>

x86/fpu: fix compile error without kernel_fpu_disabled()

In v5.18-evl-rebase, function kernel_fpu_disabled() has been removed in
commit 59f5ede3bc0f("x86/fpu: Prevent FPU state corruption"), so we will
get compile error when CONFIG_DOVETAIL is enabled:
arch/x86/kernel/fpu/core.c:931:6: error: implicit declaration of function ‘kernel_fpu_disabled’; did you mean ‘perf_pmu_disable’? [-Werror=implicit-function-declaration]
    if (kernel_fpu_disabled()) {
        ^~~~~~~~~~~~~~~~~~~
        perf_pmu_disable
cc1: all warnings being treated as errors
So, fix it.

Signed-off-by: Jamie Huang <jamienstar@163.com>
---
 arch/x86/include/asm/dovetail.h  |  18 +++-
 arch/x86/include/asm/fpu/api.h   |  25 ++++--
 arch/x86/include/asm/fpu/sched.h |  57 +++++++++---
 arch/x86/include/asm/fpu/types.h |  12 +++
 arch/x86/kernel/fpu/core.c       | 145 +++++++++++++++++++++++++++----
 arch/x86/kernel/fpu/signal.c     |  27 +++---
 arch/x86/kernel/fpu/xstate.c     |  10 ++-
 arch/x86/kernel/process_64.c     |   9 +-
 8 files changed, 252 insertions(+), 51 deletions(-)

diff --git a/arch/x86/include/asm/dovetail.h b/arch/x86/include/asm/dovetail.h
index f4b866b1a..001520262 100644
--- a/arch/x86/include/asm/dovetail.h
+++ b/arch/x86/include/asm/dovetail.h
@@ -9,14 +9,20 @@
 #if !defined(__ASSEMBLY__) && defined(CONFIG_DOVETAIL)
 
 #include <asm/io_bitmap.h>
+#include <asm/fpu/api.h>
 #include <linux/compat.h>
 
 static inline void arch_dovetail_exec_prepare(void)
-{ }
+{
+	clear_thread_flag(TIF_NEED_FPU_LOAD);
+}
 
 static inline
 void arch_dovetail_switch_prepare(bool leave_inband)
-{ }
+{
+	if (leave_inband)
+		fpu__suspend_inband();
+}
 
 static inline
 void arch_dovetail_switch_finish(bool enter_inband)
@@ -25,6 +31,14 @@ void arch_dovetail_switch_finish(bool enter_inband)
 
 	if (unlikely(ti_work & _TIF_IO_BITMAP))
 		tss_update_io_bitmap();
+
+	if (enter_inband) {
+		fpu__resume_inband();
+	} else {
+		  if (unlikely(ti_work & _TIF_NEED_FPU_LOAD &&
+				  !(current->flags & PF_KTHREAD)))
+			  switch_fpu_return();
+	}
 }
 
 /*
diff --git a/arch/x86/include/asm/fpu/api.h b/arch/x86/include/asm/fpu/api.h
index 2b3227188..86d5fed8f 100644
--- a/arch/x86/include/asm/fpu/api.h
+++ b/arch/x86/include/asm/fpu/api.h
@@ -66,20 +66,29 @@ static inline void kernel_fpu_begin(void)
  *
  * Disabling preemption also serializes against kernel_fpu_begin().
  */
-static inline void fpregs_lock(void)
+static inline unsigned long fpregs_lock(void)
 {
+	if (IS_ENABLED(CONFIG_IRQ_PIPELINE))
+		return hard_preempt_disable();
+
 	if (!IS_ENABLED(CONFIG_PREEMPT_RT))
 		local_bh_disable();
 	else
 		preempt_disable();
+
+	return 0;
 }
 
-static inline void fpregs_unlock(void)
+static inline void fpregs_unlock(unsigned long flags)
 {
-	if (!IS_ENABLED(CONFIG_PREEMPT_RT))
-		local_bh_enable();
-	else
-		preempt_enable();
+	if (IS_ENABLED(CONFIG_IRQ_PIPELINE)) {
+		hard_preempt_enable(flags);
+	} else {
+		if (!IS_ENABLED(CONFIG_PREEMPT_RT))
+			local_bh_enable();
+		else
+			preempt_enable();
+	}
 }
 
 /*
@@ -102,6 +111,10 @@ static inline void fpregs_assert_state_consistent(void) { }
  */
 extern void switch_fpu_return(void);
 
+/* For Dovetail context switching. */
+void fpu__suspend_inband(void);
+void fpu__resume_inband(void);
+
 /*
  * Query the presence of one or more xfeatures. Works on any legacy CPU as well.
  *
diff --git a/arch/x86/include/asm/fpu/sched.h b/arch/x86/include/asm/fpu/sched.h
index c485f1944..029f5a5d6 100644
--- a/arch/x86/include/asm/fpu/sched.h
+++ b/arch/x86/include/asm/fpu/sched.h
@@ -3,6 +3,7 @@
 #define _ASM_X86_FPU_SCHED_H
 
 #include <linux/sched.h>
+#include <linux/dovetail.h>
 
 #include <asm/cpufeature.h>
 #include <asm/fpu/types.h>
@@ -15,6 +16,32 @@ extern int  fpu_clone(struct task_struct *dst, unsigned long clone_flags, bool m
 		      unsigned long shstk_addr);
 extern void fpu_flush_thread(void);
 
+#ifdef CONFIG_DOVETAIL
+
+static inline void oob_fpu_set_preempt(struct fpu *fpu)
+{
+	fpu->preempted = true;
+}
+
+static inline void oob_fpu_clear_preempt(struct fpu *fpu)
+{
+	fpu->preempted = false;
+}
+
+static inline bool oob_fpu_preempted(struct fpu *old_fpu)
+{
+	return old_fpu->preempted;
+}
+
+#else
+
+static inline bool oob_fpu_preempted(struct fpu *old_fpu)
+{
+	return false;
+}
+
+#endif	/* !CONFIG_DOVETAIL */
+
 /*
  * FPU state switching for scheduling.
  *
@@ -39,22 +66,24 @@ extern void fpu_flush_thread(void);
  */
 static inline void switch_fpu_prepare(struct task_struct *old, int cpu)
 {
-	if (cpu_feature_enabled(X86_FEATURE_FPU) &&
-	    !(old->flags & (PF_KTHREAD | PF_USER_WORKER))) {
+	if (cpu_feature_enabled(X86_FEATURE_FPU)) {
 		struct fpu *old_fpu = &old->thread.fpu;
 
-		save_fpregs_to_fpstate(old_fpu);
-		/*
-		 * The save operation preserved register state, so the
-		 * fpu_fpregs_owner_ctx is still @old_fpu. Store the
-		 * current CPU number in @old_fpu, so the next return
-		 * to user space can avoid the FPU register restore
-		 * when is returns on the same CPU and still owns the
-		 * context.
-		 */
-		old_fpu->last_cpu = cpu;
-
-		trace_x86_fpu_regs_deactivated(old_fpu);
+		if (!(old->flags & (PF_KTHREAD | PF_USER_WORKER)) &&
+			!oob_fpu_preempted(old_fpu)) {
+			save_fpregs_to_fpstate(old_fpu);
+			/*
+			 * The save operation preserved register state, so the
+			 * fpu_fpregs_owner_ctx is still @old_fpu. Store the
+			 * current CPU number in @old_fpu, so the next return
+			 * to user space can avoid the FPU register restore
+			 * when is returns on the same CPU and still owns the
+			 * context.
+			 */
+			old_fpu->last_cpu = cpu;
+
+			trace_x86_fpu_regs_deactivated(old_fpu);
+		}
 	}
 }
 
diff --git a/arch/x86/include/asm/fpu/types.h b/arch/x86/include/asm/fpu/types.h
index de16862bf..121c48a09 100644
--- a/arch/x86/include/asm/fpu/types.h
+++ b/arch/x86/include/asm/fpu/types.h
@@ -462,6 +462,18 @@ struct fpu {
 	 */
 	unsigned int			last_cpu;
 
+#ifdef CONFIG_DOVETAIL
+	/*
+	 * @preempted:
+	 *
+	 * When Dovetail is enabled, this flag is set for the inband
+	 * task context saved when entering a kernel_fpu_begin/end()
+	 * section before the latter got preempted by an out-of-band
+	 * task.
+	 */
+	bool				preempted : 1;
+#endif
+
 	/*
 	 * @avx512_timestamp:
 	 *
diff --git a/arch/x86/kernel/fpu/core.c b/arch/x86/kernel/fpu/core.c
index c5316bd98..76fa5ff17 100644
--- a/arch/x86/kernel/fpu/core.c
+++ b/arch/x86/kernel/fpu/core.c
@@ -19,6 +19,7 @@
 #include <linux/hardirq.h>
 #include <linux/pkeys.h>
 #include <linux/vmalloc.h>
+#include <linux/cpuhotplug.h>
 
 #include "context.h"
 #include "internal.h"
@@ -57,6 +58,9 @@ DEFINE_PER_CPU(struct fpu *, fpu_fpregs_owner_ctx);
  */
 bool irq_fpu_usable(void)
 {
+	if (running_oob())
+		return false;
+
 	if (WARN_ON_ONCE(in_nmi()))
 		return false;
 
@@ -294,11 +298,13 @@ EXPORT_SYMBOL_GPL(fpu_enable_guest_xfd_features);
 #ifdef CONFIG_X86_64
 void fpu_update_guest_xfd(struct fpu_guest *guest_fpu, u64 xfd)
 {
-	fpregs_lock();
+	unsigned long flags;
+
+	flags = fpregs_lock();
 	guest_fpu->fpstate->xfd = xfd;
 	if (guest_fpu->fpstate->in_use)
 		xfd_update_state(guest_fpu->fpstate);
-	fpregs_unlock();
+	fpregs_unlock(flags);
 }
 EXPORT_SYMBOL_GPL(fpu_update_guest_xfd);
 
@@ -332,8 +338,9 @@ int fpu_swap_kvm_fpstate(struct fpu_guest *guest_fpu, bool enter_guest)
 	struct fpstate *guest_fps = guest_fpu->fpstate;
 	struct fpu *fpu = &current->thread.fpu;
 	struct fpstate *cur_fps = fpu->fpstate;
+	unsigned long flags;
 
-	fpregs_lock();
+	flags = fpregs_lock();
 	if (!cur_fps->is_confidential && !test_thread_flag(TIF_NEED_FPU_LOAD))
 		save_fpregs_to_fpstate(fpu);
 
@@ -363,7 +370,7 @@ int fpu_swap_kvm_fpstate(struct fpu_guest *guest_fpu, bool enter_guest)
 	}
 
 	fpregs_mark_activate();
-	fpregs_unlock();
+	fpregs_unlock(flags);
 	return 0;
 }
 EXPORT_SYMBOL_GPL(fpu_swap_kvm_fpstate);
@@ -420,11 +427,15 @@ EXPORT_SYMBOL_GPL(fpu_copy_uabi_to_guest_fpstate);
 
 void kernel_fpu_begin_mask(unsigned int kfpu_mask)
 {
+	unsigned long flags;
+
 	preempt_disable();
 
 	WARN_ON_FPU(!irq_fpu_usable());
 	WARN_ON_FPU(this_cpu_read(in_kernel_fpu));
 
+	flags = hard_cond_local_irq_save();
+
 	this_cpu_write(in_kernel_fpu, true);
 
 	if (!(current->flags & (PF_KTHREAD | PF_USER_WORKER)) &&
@@ -440,6 +451,8 @@ void kernel_fpu_begin_mask(unsigned int kfpu_mask)
 
 	if (unlikely(kfpu_mask & KFPU_387) && boot_cpu_has(X86_FEATURE_FPU))
 		asm volatile ("fninit");
+
+	hard_cond_local_irq_restore(flags);
 }
 EXPORT_SYMBOL_GPL(kernel_fpu_begin_mask);
 
@@ -458,16 +471,18 @@ EXPORT_SYMBOL_GPL(kernel_fpu_end);
  */
 void fpu_sync_fpstate(struct fpu *fpu)
 {
+	unsigned long flags;
+
 	WARN_ON_FPU(fpu != &current->thread.fpu);
 
-	fpregs_lock();
+	flags = fpregs_lock();
 	trace_x86_fpu_before_save(fpu);
 
 	if (!test_thread_flag(TIF_NEED_FPU_LOAD))
 		save_fpregs_to_fpstate(fpu);
 
 	trace_x86_fpu_after_save(fpu);
-	fpregs_unlock();
+	fpregs_unlock(flags);
 }
 
 static inline unsigned int init_fpstate_copy_size(void)
@@ -586,6 +601,7 @@ int fpu_clone(struct task_struct *dst, unsigned long clone_flags, bool minimal,
 {
 	struct fpu *src_fpu = &current->thread.fpu;
 	struct fpu *dst_fpu = &dst->thread.fpu;
+	unsigned long flags;
 
 	/* The new task's FPU state cannot be valid in the hardware. */
 	dst_fpu->last_cpu = -1;
@@ -627,11 +643,16 @@ int fpu_clone(struct task_struct *dst, unsigned long clone_flags, bool minimal,
 	 * Do not use memcpy() when TIF_NEED_FPU_LOAD is set because
 	 * copying is not valid when current uses non-default states.
 	 */
-	fpregs_lock();
+	flags = fpregs_lock();
 	if (test_thread_flag(TIF_NEED_FPU_LOAD))
 		fpregs_restore_userregs();
 	save_fpregs_to_fpstate(dst_fpu);
-	fpregs_unlock();
+	fpregs_unlock(flags);
+	/*
+	 * Dovetail: this is ok to access the current fpu perm bits of
+	 * current for reading only, even in case of preemption by an
+	 * oob thread.
+	 */
 	if (!(clone_flags & CLONE_THREAD))
 		fpu_inherit_perms(dst_fpu);
 
@@ -675,7 +696,9 @@ void fpu_thread_struct_whitelist(unsigned long *offset, unsigned long *size)
  */
 void fpu__drop(struct fpu *fpu)
 {
-	preempt_disable();
+	unsigned long flags;
+
+	flags = hard_preempt_disable();
 
 	if (fpu == &current->thread.fpu) {
 		/* Ignore delayed exceptions from user space */
@@ -687,7 +710,7 @@ void fpu__drop(struct fpu *fpu)
 
 	trace_x86_fpu_dropped(fpu);
 
-	preempt_enable();
+	hard_preempt_enable(flags);
 }
 
 /*
@@ -712,8 +735,9 @@ static inline void restore_fpregs_from_init_fpstate(u64 features_mask)
 static void fpu_reset_fpregs(void)
 {
 	struct fpu *fpu = &current->thread.fpu;
+	unsigned long flags;
 
-	fpregs_lock();
+	flags = fpregs_lock();
 	__fpu_invalidate_fpregs_state(fpu);
 	/*
 	 * This does not change the actual hardware registers. It just
@@ -730,7 +754,7 @@ static void fpu_reset_fpregs(void)
 	 */
 	memcpy(&fpu->fpstate->regs, &init_fpstate.regs, init_fpstate_copy_size());
 	set_thread_flag(TIF_NEED_FPU_LOAD);
-	fpregs_unlock();
+	fpregs_unlock(flags);
 }
 
 /*
@@ -740,12 +764,14 @@ static void fpu_reset_fpregs(void)
  */
 void fpu__clear_user_states(struct fpu *fpu)
 {
+	unsigned long flags;
+
 	WARN_ON_FPU(fpu != &current->thread.fpu);
 
-	fpregs_lock();
+	flags = fpregs_lock();
 	if (!cpu_feature_enabled(X86_FEATURE_FPU)) {
 		fpu_reset_fpregs();
-		fpregs_unlock();
+		fpregs_unlock(flags);
 		return;
 	}
 
@@ -768,7 +794,7 @@ void fpu__clear_user_states(struct fpu *fpu)
 	 * current's FPU is marked active.
 	 */
 	fpregs_mark_activate();
-	fpregs_unlock();
+	fpregs_unlock(flags);
 }
 
 void fpu_flush_thread(void)
@@ -781,10 +807,14 @@ void fpu_flush_thread(void)
  */
 void switch_fpu_return(void)
 {
+	unsigned long flags;
+
 	if (!static_cpu_has(X86_FEATURE_FPU))
 		return;
 
+	flags = hard_cond_local_irq_save();
 	fpregs_restore_userregs();
+	hard_cond_local_irq_restore(flags);
 }
 EXPORT_SYMBOL_GPL(switch_fpu_return);
 
@@ -919,3 +949,88 @@ noinstr void fpu_idle_fpregs(void)
 		__this_cpu_write(fpu_fpregs_owner_ctx, NULL);
 	}
 }
+
+#ifdef CONFIG_DOVETAIL
+
+/*
+ * Holds the in-kernel fpu state when preempted by a task running on
+ * the out-of-band stage.
+ */
+static DEFINE_PER_CPU(struct fpu *, in_kernel_fpstate);
+
+static int fpu__init_kernel_fpstate(unsigned int cpu)
+{
+	struct fpu *fpu;
+	int fpu_size;
+
+	fpu_size = sizeof(current->thread.fpu);
+	fpu_size -= sizeof(current->thread.fpu.__fpstate.regs);
+	fpu_size += fpu_kernel_cfg.default_size;
+
+	/*
+	 * We need to save kernel fpu states exclusively (statically
+	 * sized). See fpu__suspend_inband().
+	 */
+	fpu = kzalloc(fpu_size, GFP_KERNEL);
+	if (fpu == NULL)
+		return -ENOMEM;
+
+	fpu->last_cpu = -1;
+	fpstate_reset(fpu);
+	this_cpu_write(in_kernel_fpstate, fpu);
+	memcpy(&fpu->fpstate->regs, &init_fpstate.regs,
+		init_fpstate_copy_size());
+
+	return 0;
+}
+
+static int fpu__drop_kernel_fpstate(unsigned int cpu)
+{
+	struct fpu *fpu = this_cpu_read(in_kernel_fpstate);
+
+	kfree(fpu);
+
+	return 0;
+}
+
+void fpu__suspend_inband(void)
+{
+	struct fpu *kfpu = this_cpu_read(in_kernel_fpstate);
+	struct task_struct *tsk = current;
+
+	/*
+	 * If in_kernel_fpu is set, we are dealing with the preemption
+	 * of an inband kernel context currently using the fpu by a
+	 * thread which resumes on the oob stage.
+	 */
+	if (this_cpu_read(in_kernel_fpu)) {
+		save_fpregs_to_fpstate(kfpu);
+		__cpu_invalidate_fpregs_state();
+		oob_fpu_set_preempt(&tsk->thread.fpu);
+	}
+}
+
+void fpu__resume_inband(void)
+{
+	struct fpu *kfpu = this_cpu_read(in_kernel_fpstate);
+	struct task_struct *tsk = current;
+
+	if (oob_fpu_preempted(&tsk->thread.fpu)) {
+		restore_fpregs_from_fpstate(kfpu->fpstate, XFEATURE_MASK_FPSTATE);
+		__cpu_invalidate_fpregs_state();
+		oob_fpu_clear_preempt(&tsk->thread.fpu);
+	} else if (!(tsk->flags & PF_KTHREAD) &&
+		test_thread_flag(TIF_NEED_FPU_LOAD))
+		switch_fpu_return();
+}
+
+static int __init fpu__init_dovetail(void)
+{
+	cpuhp_setup_state(CPUHP_AP_ONLINE_DYN,
+			"platform/x86/dovetail:online",
+			fpu__init_kernel_fpstate, fpu__drop_kernel_fpstate);
+	return 0;
+}
+core_initcall(fpu__init_dovetail);
+
+#endif
diff --git a/arch/x86/kernel/fpu/signal.c b/arch/x86/kernel/fpu/signal.c
index 8abe60919..1784ad9c1 100644
--- a/arch/x86/kernel/fpu/signal.c
+++ b/arch/x86/kernel/fpu/signal.c
@@ -68,15 +68,17 @@ static inline bool check_xstate_in_sigframe(struct fxregs_state __user *fxbuf,
  */
 static inline bool save_fsave_header(struct task_struct *tsk, void __user *buf)
 {
+	unsigned long flags;
+
 	if (use_fxsr()) {
 		struct xregs_state *xsave = &tsk->thread.fpu.fpstate->regs.xsave;
 		struct user_i387_ia32_struct env;
 		struct _fpstate_32 __user *fp = buf;
 
-		fpregs_lock();
+		flags = fpregs_lock();
 		if (!test_thread_flag(TIF_NEED_FPU_LOAD))
 			fxsave(&tsk->thread.fpu.fpstate->regs.fxsave);
-		fpregs_unlock();
+		fpregs_unlock(flags);
 
 		convert_from_fxsr(&env, tsk);
 
@@ -182,6 +184,7 @@ bool copy_fpstate_to_sigframe(void __user *buf, void __user *buf_fx, int size, u
 	struct task_struct *tsk = current;
 	struct fpstate *fpstate = tsk->thread.fpu.fpstate;
 	bool ia32_fxstate = (buf != buf_fx);
+	unsigned long flags;
 	int ret;
 
 	ia32_fxstate &= (IS_ENABLED(CONFIG_X86_32) ||
@@ -215,14 +218,14 @@ bool copy_fpstate_to_sigframe(void __user *buf, void __user *buf_fx, int size, u
 	 * userland's stack frame which will likely succeed. If it does not,
 	 * resolve the fault in the user memory and try again.
 	 */
-	fpregs_lock();
+	flags = fpregs_lock();
 	if (test_thread_flag(TIF_NEED_FPU_LOAD))
 		fpregs_restore_userregs();
 
 	pagefault_disable();
 	ret = copy_fpregs_to_sigframe(buf_fx, pkru);
 	pagefault_enable();
-	fpregs_unlock();
+	fpregs_unlock(flags);
 
 	if (ret) {
 		if (!__clear_user(buf_fx, fpstate->user_size))
@@ -269,12 +272,13 @@ static int __restore_fpregs_from_user(void __user *buf, u64 ufeatures,
 static bool restore_fpregs_from_user(void __user *buf, u64 xrestore, bool fx_only)
 {
 	struct fpu *fpu = &current->thread.fpu;
+	unsigned long flags;
 	int ret;
 
 	/* Restore enabled features only. */
 	xrestore &= fpu->fpstate->user_xfeatures;
 retry:
-	fpregs_lock();
+	flags = fpregs_lock();
 	/* Ensure that XFD is up to date */
 	xfd_update_state(fpu->fpstate);
 	pagefault_disable();
@@ -296,7 +300,7 @@ static bool restore_fpregs_from_user(void __user *buf, u64 xrestore, bool fx_onl
 		 */
 		if (test_thread_flag(TIF_NEED_FPU_LOAD))
 			__cpu_invalidate_fpregs_state();
-		fpregs_unlock();
+		fpregs_unlock(flags);
 
 		/* Try to handle #PF, but anything else is fatal. */
 		if (ret != X86_TRAP_PF)
@@ -320,7 +324,7 @@ static bool restore_fpregs_from_user(void __user *buf, u64 xrestore, bool fx_onl
 		os_xrstor_supervisor(fpu->fpstate);
 
 	fpregs_mark_activate();
-	fpregs_unlock();
+	fpregs_unlock(flags);
 	return true;
 }
 
@@ -333,6 +337,7 @@ static bool __fpu_restore_sig(void __user *buf, void __user *buf_fx,
 	bool success, fx_only = false;
 	union fpregs_state *fpregs;
 	u64 user_xfeatures = 0;
+	unsigned long flags;
 
 	if (use_xsave()) {
 		struct _fpx_sw_bytes fx_sw_user;
@@ -365,7 +370,7 @@ static bool __fpu_restore_sig(void __user *buf, void __user *buf_fx,
 	 * to be loaded again on return to userland (overriding last_cpu avoids
 	 * the optimisation).
 	 */
-	fpregs_lock();
+	flags = fpregs_lock();
 	if (!test_thread_flag(TIF_NEED_FPU_LOAD)) {
 		/*
 		 * If supervisor states are available then save the
@@ -381,7 +386,7 @@ static bool __fpu_restore_sig(void __user *buf, void __user *buf_fx,
 	}
 	__fpu_invalidate_fpregs_state(fpu);
 	__cpu_invalidate_fpregs_state();
-	fpregs_unlock();
+	fpregs_unlock(flags);
 
 	fpregs = &fpu->fpstate->regs;
 	if (use_xsave() && !fx_only) {
@@ -409,7 +414,7 @@ static bool __fpu_restore_sig(void __user *buf, void __user *buf_fx,
 	/* Fold the legacy FP storage */
 	convert_to_fxsr(&fpregs->fxsave, &env);
 
-	fpregs_lock();
+	flags = fpregs_lock();
 	if (use_xsave()) {
 		/*
 		 * Remove all UABI feature bits not set in user_xfeatures
@@ -432,7 +437,7 @@ static bool __fpu_restore_sig(void __user *buf, void __user *buf_fx,
 	if (likely(success))
 		fpregs_mark_activate();
 
-	fpregs_unlock();
+	fpregs_unlock(flags);
 	return success;
 }
 
diff --git a/arch/x86/kernel/fpu/xstate.c b/arch/x86/kernel/fpu/xstate.c
index 22abb5ee0..ff8a175e7 100644
--- a/arch/x86/kernel/fpu/xstate.c
+++ b/arch/x86/kernel/fpu/xstate.c
@@ -1022,6 +1022,7 @@ int arch_set_user_pkey_access(struct task_struct *tsk, int pkey,
 			      unsigned long init_val)
 {
 	u32 old_pkru, new_pkru_bits = 0;
+	unsigned long flags;
 	int pkey_shift;
 
 	/*
@@ -1049,6 +1050,8 @@ int arch_set_user_pkey_access(struct task_struct *tsk, int pkey,
 	pkey_shift = pkey * PKRU_BITS_PER_PKEY;
 	new_pkru_bits <<= pkey_shift;
 
+	flags = hard_cond_local_irq_save();
+
 	/* Get old PKRU and mask off any old bits in place: */
 	old_pkru = read_pkru();
 	old_pkru &= ~((PKRU_AD_BIT|PKRU_WD_BIT) << pkey_shift);
@@ -1056,6 +1059,8 @@ int arch_set_user_pkey_access(struct task_struct *tsk, int pkey,
 	/* Write old part along with new part: */
 	write_pkru(old_pkru | new_pkru_bits);
 
+	hard_cond_local_irq_restore(flags);
+
 	return 0;
 }
 #endif /* ! CONFIG_ARCH_HAS_PKEYS */
@@ -1523,6 +1528,7 @@ static int fpstate_realloc(u64 xfeatures, unsigned int ksize,
 	struct fpstate *curfps, *newfps = NULL;
 	unsigned int fpsize;
 	bool in_use;
+	unsigned long flags;
 
 	fpsize = ksize + ALIGN(offsetof(struct fpstate, regs), 64);
 
@@ -1550,7 +1556,7 @@ static int fpstate_realloc(u64 xfeatures, unsigned int ksize,
 		guest_fpu->uabi_size = usize;
 	}
 
-	fpregs_lock();
+	flags = fpregs_lock();
 	/*
 	 * If @curfps is in use, ensure that the current state is in the
 	 * registers before swapping fpstate as that might invalidate it
@@ -1577,7 +1583,7 @@ static int fpstate_realloc(u64 xfeatures, unsigned int ksize,
 
 	if (in_use)
 		xfd_update_state(fpu->fpstate);
-	fpregs_unlock();
+	fpregs_unlock(flags);
 
 	/* Only free valloc'ed state */
 	if (curfps && curfps->is_valloc)
diff --git a/arch/x86/kernel/process_64.c b/arch/x86/kernel/process_64.c
index 9118b5742..a5f819e69 100644
--- a/arch/x86/kernel/process_64.c
+++ b/arch/x86/kernel/process_64.c
@@ -613,8 +613,15 @@ __switch_to(struct task_struct *prev_p, struct task_struct *next_p)
 	struct thread_struct *next = &next_p->thread;
 	int cpu = smp_processor_id();
 
+	/*
+	 * Dovetail: Switching context on the out-of-band stage is
+	 * legit, and we may have preempted an in-band (soft)irq
+	 * handler earlier. Since oob handlers never switch stack,
+	 * make sure to restrict the following test to in-band
+	 * callers.
+	 */
 	WARN_ON_ONCE(IS_ENABLED(CONFIG_DEBUG_ENTRY) &&
-		     this_cpu_read(pcpu_hot.hardirq_stack_inuse));
+		     running_inband() && this_cpu_read(pcpu_hot.hardirq_stack_inuse));
 
 	WARN_ON_ONCE(dovetail_debug() && !hard_irqs_disabled());
 
-- 
2.30.2

